In [274]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
import numpy as np
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [275]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [276]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [277]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [278]:
# convert data type to int64
df_market_data['price_change_percentage_24h'] = df_market_data['price_change_percentage_24h'].astype(np.int64)
df_market_data['price_change_percentage_7d'] = df_market_data['price_change_percentage_7d'].astype(np.int64)
df_market_data['price_change_percentage_14d'] = df_market_data['price_change_percentage_14d'].astype(np.int64)
df_market_data['price_change_percentage_30d'] = df_market_data['price_change_percentage_30d'].astype(np.int64)
df_market_data['price_change_percentage_60d'] = df_market_data['price_change_percentage_60d'].astype(np.int64)
df_market_data['price_change_percentage_200d'] = df_market_data['price_change_percentage_200d'].astype(np.int64)
df_market_data['price_change_percentage_1y'] = df_market_data['price_change_percentage_1y'].astype(np.int64)

# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
market_data_scaled = StandardScaler().fit_transform(df_market_data[["price_change_percentage_24h", "price_change_percentage_7d", "price_change_percentage_14d", "price_change_percentage_30d", "price_change_percentage_60d", "price_change_percentage_200d", "price_change_percentage_1y"]])

In [280]:
# Create a DataFrame with the scaled data
scaled_df = pd.DataFrame(market_data_scaled, columns=['price_change_percentage_24h', 'price_change_percentage_7d', 'price_change_percentage_14d', 'price_change_percentage_30d', 'price_change_percentage_60d', 'price_change_percentage_200d', 'price_change_percentage_1y'])

# Copy the crypto names from the original data
coin_id = ["bitcoin", "ethereum", "tether", "ripple", "bitcoin-cash", "binancecoin", "chainlink", "cardano", "litecoin", "bitcoin-cash-sv", "crypto-com-chain", \
           "usd-coin", "eos", "monero", "tron", "tezos", "okb", "stellar", "cosmos", "cdai", "neo", "wrapped-bitcoin", "leo-token", "huobi-token", \
           "nem", "binance-usd", "iota", "vechain", "zcash", "theta-token", "dash", "ethereum-classic", "ethlend", "maker", "havven", "omisego", \
           "celsius-degree-token", "ontology", "ftx-token", "true-usd", "digibyte"]
coin_id_df = pd.DataFrame(coin_id)

# concatenate the columns?
scaled_df = pd.concat([coin_id_df, scaled_df], axis=1)

# Set the coin_id column as index
scaled_df.set_index(0, inplace=True)
scaled_df = scaled_df.rename_axis('coin_id')

# Display sample data
print(scaled_df)

                      price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.524157                    0.446116   
ethereum                                 0.110879                    0.944955   
tether                                   0.110879                   -0.717842   
ripple                                   0.110879                   -0.717842   
bitcoin-cash                             0.937435                    2.108913   
binancecoin                              0.937435                    1.277515   
chainlink                                0.110879                    2.607752   
cardano                                  0.110879                    1.443794   
litecoin                                 0.110879                    0.279837   
bitcoin-cash-sv                          0.110879                   -0.219003   
crypto-com-chain            

---

### Find the Best Value for k Using the Original Data.

In [281]:
# Create a list with the number of k-values from 1 to 11
k_values = [k for k in range(1,11)]

In [282]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for k in k_values:
    model = KMeans(n_clusters=k, random_state=0)
    model.fit(scaled_df)
    inertia.append(model.inertia_)


c:\Users\rache\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\rache\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\rache\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\rache\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is k

In [283]:
# Create a dictionary with the data to plot the Elbow curve
elbow = {'k': k_values, 'inertia': inertia}

# Create a DataFrame with the data to plot the Elbow curve
elbow_df = pd.DataFrame(elbow)


In [284]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
k_means = elbow_df.hvplot.line(x='k', y='inertia', title='Elbow Curve for K-means Clustering', xlabel='Number of Clusters (k)', ylabel='Inertia')
k_means

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [285]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=0)

In [286]:
# Fit the K-Means model using the scaled data
model.fit(scaled_df)

c:\Users\rache\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=0)

In [287]:
# Predict the clusters to group the cryptocurrencies using the scaled data
k_4 = model.predict(scaled_df)

# copy the preprocessed data
predictions_df = scaled_df.copy()

# somethin or another
predictions_df['predicted_clusters'] = k_4

# Print the resulting array of cluster values.
k_4

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 1, 2, 0, 0, 3, 0, 0, 0, 0])

In [304]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

# Create the scatter plot

predictions_plot = predictions_df.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    c="predicted_clusters",
    hover_cols=["coin_id"]
    )

# Display the scatter plot
predictions_plot



:Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,predicted_clusters,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [289]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [290]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
pca_data = pca.fit_transform(scaled_df)
# View the first five rows of the DataFrame. 
pca_data = pd.DataFrame(pca_data, columns=['PCA1', 'PCA2', 'PCA3'])
pca_data.head()

PCA1      PCA2      PCA3
0 -0.634498  0.776989  0.428682
1 -0.437697  0.363349  0.957615
2 -0.461238 -0.141336 -0.669299
3 -0.529301 -0.164210 -0.450918
4 -1.134028  1.877951  2.035012

In [291]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.38092249, 0.34764686, 0.17680666])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total explained variance is 0.38092 + 0.34765 + 0.17681, or 0.90538 (91%). 

In [292]:
# Create a new DataFrame with the PCA data.
pca_data = pd.DataFrame(pca_data, columns=['PCA1', 'PCA2', 'PCA3'])

# Copy the crypto names from the original data
coin_id = ["bitcoin", "ethereum", "tether", "ripple", "bitcoin-cash", "binancecoin", "chainlink", "cardano", "litecoin", "bitcoin-cash-sv", "crypto-com-chain", \
           "usd-coin", "eos", "monero", "tron", "tezos", "okb", "stellar", "cosmos", "cdai", "neo", "wrapped-bitcoin", "leo-token", "huobi-token", \
           "nem", "binance-usd", "iota", "vechain", "zcash", "theta-token", "dash", "ethereum-classic", "ethlend", "maker", "havven", "omisego", \
           "celsius-degree-token", "ontology", "ftx-token", "true-usd", "digibyte"]
coin_id_df = pd.DataFrame(coin_id)

# concatenate the columns?
pca_data = pd.concat([coin_id_df, pca_data], axis=1)

# Set the coinid column as index
pca_data.set_index(0, inplace=True)
pca_data = pca_data.rename_axis('coin_id')

# Display sample data
pca_data.head()

PCA1      PCA2      PCA3
coin_id                                   
bitcoin      -0.634498  0.776989  0.428682
ethereum     -0.437697  0.363349  0.957615
tether       -0.461238 -0.141336 -0.669299
ripple       -0.529301 -0.164210 -0.450918
bitcoin-cash -1.134028  1.877951  2.035012

---

### Find the Best Value for k Using the PCA Data

In [293]:
# Create a list with the number of k-values from 1 to 11
pca_k_means = list(range(1,11))

In [294]:
# Create an empty list to store the inertia values
pca_inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list

for i in pca_k_means:
    k_model = KMeans(n_clusters=i, random_state=1, n_init=10)
    k_model.fit(pca_data)
    pca_inertia.append(k_model.inertia_)

c:\Users\rache\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\rache\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\rache\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\rache\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is k

In [295]:
# Create a dictionary with the data to plot the Elbow curve
pca_elbow = {"pca_k_means": pca_k_means, "pca_inertia": pca_inertia}

# Create a DataFrame with the data to plot the Elbow curve
pca_elbow_df = pd.DataFrame(pca_elbow)

# check dataframe
pca_elbow_df.head()

pca_k_means  pca_inertia
0            1   259.842916
1            2   165.778181
2            3    92.826943
3            4    49.289374
4            5    38.137694

In [296]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
pca_elbow = pca_elbow_df.hvplot.line(
    x="pca_k_means",
    y="pca_inertia",
    title="PCA Elbow Curve",
    color='blue',
    xticks=pca_k_means
)
pca_elbow

:Curve   [pca_k_means]   (pca_inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No.

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [297]:
# Initialize the K-Means model using the best value for k
pca_model = KMeans(n_clusters=4)

In [298]:
# Fit the K-Means model using the PCA data
pca_model.fit(pca_data)

c:\Users\rache\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4)

In [299]:
# Predict the clusters to group the cryptocurrencies using the PCA data
pca_predictions = pca_model.predict(pca_data)

# Print the resulting array of cluster values.
print(pca_predictions)


[3 3 0 0 3 3 3 3 3 0 0 0 0 3 0 3 0 0 3 0 0 3 0 0 0 0 0 0 3 0 0 0 1 3 0 0 2
 0 0 0 0]


In [300]:
# Create a copy of the DataFrame with the PCA data
pca_predictions_df = pca_data.copy()

# Add a new column to the DataFrame with the predicted clusters
pca_predictions_df['pca_predictions'] = pca_predictions

# Display sample data
pca_predictions_df.head()


PCA1      PCA2      PCA3  pca_predictions
coin_id                                                    
bitcoin      -0.634498  0.776989  0.428682                3
ethereum     -0.437697  0.363349  0.957615                3
tether       -0.461238 -0.141336 -0.669299                0
ripple       -0.529301 -0.164210 -0.450918                0
bitcoin-cash -1.134028  1.877951  2.035012                3

In [301]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

pca_predictions_plot = pca_predictions_df.hvplot.scatter(
    x='PCA1', 
    y='PCA2',
    title="PCA Clustering",
    by='pca_predictions',
    hover_cols='coin_id'
)
pca_predictions_plot

:NdOverlay   [pca_predictions]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [302]:
# Composite plot to contrast the Elbow curves'

elbow_composite = (k_means + pca_elbow).cols(1)
elbow_composite

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [pca_k_means]   (pca_inertia)

In [305]:
# Composite plot to contrast the clusters
cluster_composite = (predictions_plot + pca_predictions_plot).cols(1)
cluster_composite

:Layout
   .Scatter.I   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,predicted_clusters,coin_id)
   .NdOverlay.I :NdOverlay   [pca_predictions]
      :Scatter   [PCA1]   (PCA2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Both algorithms predicted that clusters would be near 0. K-Means shows a clear line slightly greater than 0 and captures more of the negative integer outliers. PCA shows two different colored blobs slightly less than 0 and captures more of the positive integer outliers. K-Means clusters provide a clear graph with high readability, with minimal loss of information or accuracy.